python/semantic_kernel/orchestration/kernel_function.py

```diff
@@ -367,24 +389,49 @@ def invoke(
             context.variables.update(input)
 
         try:
-            loop = asyncio.get_running_loop() if asyncio.get_event_loop().is_running() else None
+            loop = asyncio.get_event_loop()
         except RuntimeError:
-            loop = None
-
-        if loop and loop.is_running():
-
-            def run_coroutine():
-                if self.is_semantic:
-                    return self._invoke_semantic(context, settings)
-                else:
-                    return self._invoke_native(context)
+            # Create a new event loop if one isn't already present
+            loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(loop)
 
-            return self.run_async_in_executor(run_coroutine)
+        if loop.is_running():
+            # Schedule the coroutine on the existing event loop
+            # If we get here, it's because there's already an event loop running
+            # We can't block and  wait for the result, so we need to schedule it
+            # and then return it.
+            # The caller of this task should await the result
+            task = asyncio.create_task(
+                self.invoke_async(
+                    input=input,
+                    variables=variables,
+                    context=context,
+                    memory=memory,
+                    settings=settings,
+                    log=log,
+                )
+            )
+            return task
```

Comment:
this is not a finished task, shouldn't it just run asyncio.run() and then the actual result is returned?

Reply:
Simply calling `asyncio.run()` is the reason the execution hangs. A Jupyter notebook already has a running event loop, if you try to fire off an `asyncio.run()`, it will create another event loop, and we get into trouble. I've already converted the Jupyter notebook code to use `invoke_async` where it was previously using `invoke`, aside from the methods that are hitting the overriden `__call__` method in the `kernel_function`. I think we should move away from supporting sync invocations, but we can discuss offline.